test/hotspot/jtreg/compiler/loopopts/TestUnswitchPredicateCloning.java

```diff
@@ -0,0 +1,185 @@
+                   IRNode.AUTO_VECTORIZATION_CHECK_PARSE_PREDICATE, "3" },
+        phase = CompilePhase.BEFORE_LOOP_UNSWITCHING)
+    // Since we know that predication happens after unswitching, we can test the
+    // predicate cloning before predication, such that the useless, killed predicates
+    // have already been removed in the beautify loop phase.
+    @IR(counts = { IRNode.LOOP_PARSE_PREDICATE, "4",
+                   IRNode.PROFILED_LOOP_PARSE_PREDICATE, "4",
+                   IRNode.LOOP_LIMIT_CHECK_PARSE_PREDICATE, "3",
+                   IRNode.AUTO_VECTORIZATION_CHECK_PARSE_PREDICATE, "4" },
+        phase = CompilePhase.BEFORE_LOOP_PREDICATION_IC)
+    // Check that opaque template assertion predicated are added in loop predication
+    // even if loop predication only happens after loop unswitching.
+    @IR(failOn = { IRNode.OPAQUE_TEMPLATE_ASSERTION_PREDICATE },
+        phase = CompilePhase.AFTER_LOOP_UNSWITCHING)
+    @IR(counts = { IRNode.OPAQUE_TEMPLATE_ASSERTION_PREDICATE, "2" },
+        phase = CompilePhase.AFTER_LOOP_PREDICATION_RC)
+    static int testUnswitchingBeforePredication(int j) {
+        int zero = 34;
+        int limit = 2;
+
+        // Ensure zero == 0 is only known after CCP
+        for (; limit < 4; limit *= 2) {
+        }
+        for (int i = 2; i < limit; i++) {
+            zero = 0;
+        }
+
+        int[] arr = getArr();
+        int res = 0;
+        for (int i = 0; i < arr.length; i++) {
+            // Trigger unswitching only after CCP
+            if (zero == 0) {
+                // Trigger range check after loop unswitching
+                res += arr[j];
+            } else {
+                res += arr[i];
+            }
+        }
+
+        return res;
+    }
+
+    @Test
+    // Check that Loop Unswitching doubled the number of parse and tempalte
+    // assertion predicates. Again, the Loop Limit Check Parse Predicate
+    // remains at the Loop Selector since this is a counted loop.
+    @IR(failOn = { IRNode.OPAQUE_TEMPLATE_ASSERTION_PREDICATE },
+        phase = CompilePhase.BEFORE_LOOP_PREDICATION_RC)
+    @IR(counts = { IRNode.OPAQUE_TEMPLATE_ASSERTION_PREDICATE, "2",
+                   IRNode.LOOP_PARSE_PREDICATE, "1",
+                   IRNode.PROFILED_LOOP_PARSE_PREDICATE, "1",
+                   IRNode.LOOP_LIMIT_CHECK_PARSE_PREDICATE, "1",
+                   IRNode.AUTO_VECTORIZATION_CHECK_PARSE_PREDICATE, "1" },
+        phase = CompilePhase.BEFORE_LOOP_UNSWITCHING)
+    // After loop unswitching and after removing the killed predicates.
+    @IR(counts = { IRNode.OPAQUE_TEMPLATE_ASSERTION_PREDICATE, "4",
+                   IRNode.LOOP_PARSE_PREDICATE, "2",
+                   IRNode.PROFILED_LOOP_PARSE_PREDICATE, "2",
+                   IRNode.LOOP_LIMIT_CHECK_PARSE_PREDICATE, "1",
+                   IRNode.AUTO_VECTORIZATION_CHECK_PARSE_PREDICATE, "2" },
+        phase = CompilePhase.BEFORE_LOOP_PREDICATION_IC)
```

Comment:
This suggests that we apply another round of Loop Predication later? It might be confused with the first application of Loop Predication. Could we also match on `PHASEIDEALLOOP_ITERATIONS` instead? Would that work?